From patchwork Tue Apr 24 05:46:50 2012
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
Subject: [synaptics] Reset touch state on DeviceOff
Date: Tue, 24 Apr 2012 05:46:50 -0000
From: Peter Hutterer <peter.hutterer@who-t.net>
X-Patchwork-Id: 10055
Message-Id: <20120424054650.GA1803@yabbi.bne.redhat.com>
To: "X.Org Devel List" <xorg-devel@lists.freedesktop.org>
Cc: Chase Douglas <chase.douglas@canonical.com>

Don't leave touches lingering around during suspend.

Test case:
1) leave finger on touchpad
2) xinput set-prop "SynPS/2 Synaptics TouchPad" "Device Enabled" 0
3) lift fingers
4) xinput set-prop "SynPS/2 Synaptics TouchPad" "Device Enabled" 1

https://bugzilla.redhat.com/show_bug.cgi?id=814972

Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>

---
src/eventcomm.c |    1 +
 src/synaptics.c |    1 +
 2 files changed, 2 insertions(+)

diff --git a/src/eventcomm.c b/src/eventcomm.c
index 9d1233c..5707e38 100644
--- a/src/eventcomm.c
+++ b/src/eventcomm.c
@@ -128,6 +128,7 @@ UninitializeTouch(InputInfoPtr pInfo)
 
     mtdev_close(proto_data->mtdev);
     proto_data->mtdev = NULL;
+    proto_data->num_touches = 0;
 }
 
 static void
diff --git a/src/synaptics.c b/src/synaptics.c
index 853bfa8..d2fe960 100644
--- a/src/synaptics.c
+++ b/src/synaptics.c
@@ -1086,6 +1086,7 @@ DeviceOff(DeviceIntPtr dev)
     if (pInfo->fd != -1) {
 	TimerCancel(priv->timer);
 	xf86RemoveEnabledDevice(pInfo);
+        SynapticsResetTouchHwState(priv->hwState);
         if (priv->proto_ops->DeviceOffHook &&
             !priv->proto_ops->DeviceOffHook(pInfo))
             rc = !Success;
